### PR TITLE
Client and Server Plugins and Testing

### PR DESCRIPTION
**Pull request by loosebazooka**

### CODE REVIEW

**Comment by loosebazooka** [comment]
@patflynn @meltsufin 

tests for the plugin, see #1 and #2 first

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+out
+*.iml
+*.ipr
+*.iws
+.idea
+.gradle
+/.settings
+/.classpath
+/.project
+gradle.properties
```
Comment (patflynn) [inline .gitignore:11]: you want to ignore gradle.properties?

File: appveyor.yml [hunk 2]
```diff
@@ -0,0 +1,26 @@
+version: "{branch} {build}"
```
Comment (patflynn) [inline appveyor.yml:1]: why travis and appveyor?

File: src/main/java/com/google/cloud/tools/gradle/endpoints/client/EndpointsClientPlugin.java [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+/**
+ * Plugin definition for Endpoints Clients (will generate source when applied)
+ *
+ * All tasks are hidden. Configuration of source discovery docs is from two ways:
+ * 1. specify the location of the discovery doc with the extension
+ * endpointsClient {
```
Comment (patflynn) [inline src/main/java/com/google/cloud/tools/gradle/endpoints/client/EndpointsClientPlugin.java:44]: {@ code..}

File: src/main/java/com/google/cloud/tools/gradle/endpoints/client/EndpointsClientPlugin.java [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+ *   discoveryDocs = [file(path/to/xyz.discovery)]
```
Comment (patflynn) [inline src/main/java/com/google/cloud/tools/gradle/endpoints/client/EndpointsClientPlugin.java:45]: can it also be a url?

File: src/main/java/com/google/cloud/tools/gradle/endpoints/client/EndpointsClientPlugin.java [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+import org.gradle.api.tasks.SourceSet;
+import org.gradle.api.tasks.SourceSetContainer;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+/**
+ * Plugin definition for Endpoints Clients (will generate source when applied)
```
Comment (patflynn) [inline src/main/java/com/google/cloud/tools/gradle/endpoints/client/EndpointsClientPlugin.java]: where does it generate src?

File: src/main/java/com/google/cloud/tools/gradle/endpoints/client/EndpointsClientPlugin.java [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+            new Action<GenerateClientLibrarySourceTask>() {
+              @Override
+              public void execute(final GenerateClientLibrarySourceTask genClientLibSrc) {
+                genClientLibSrc.setDescription("_internal");
+                genClientLibSrc.dependsOn(GENERATE_CLIENT_LIBRARY_TASK);
+
+                project.afterEvaluate(new Action<Project>() {
+                  @Override
+                  public void execute(Project project) {
+                    genClientLibSrc.setClientLibDir(extension.getClientLibDir());
```
Comment (patflynn) [inline src/main/java/com/google/cloud/tools/gradle/endpoints/client/EndpointsClientPlugin.java:172]: what's the difference between the clientLibDir and the generatedSrcDir?

File: src/main/java/com/google/cloud/tools/gradle/endpoints/client/task/GetDiscoveryDocsFromDependenciesTask.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+import org.gradle.api.DefaultTask;
+import org.gradle.api.tasks.InputFiles;
+import org.gradle.api.tasks.OutputDirectory;
+import org.gradle.api.tasks.TaskAction;
+
+import java.io.File;
+import java.util.Collection;
+import java.util.HashMap;
+
+public class GetDiscoveryDocsFromDependenciesTask extends DefaultTask {
```
Comment (patflynn) [inline src/main/java/com/google/cloud/tools/gradle/endpoints/client/task/GetDiscoveryDocsFromDependenciesTask.java]: javadoc

**Review by patflynn — COMMENTED** [review]
This looks pretty good. the ability to specify a backend module is interesting.  Ideally we'd send these PRs out in smaller chunks.

File: .gitignore (hunk 1 above)
Comment (loosebazooka) [reply .gitignore:11]: Yeah, typically I use it as a local properties file

File: appveyor.yml (hunk 2 above)
Comment (loosebazooka) [reply appveyor.yml:1]: windows for appveyor.

File: src/main/java/com/google/cloud/tools/gradle/endpoints/client/EndpointsClientPlugin.java (hunk 4 above)
Comment (loosebazooka) [reply src/main/java/com/google/cloud/tools/gradle/endpoints/client/EndpointsClientPlugin.java:45]: no, should it be?

File: src/main/java/com/google/cloud/tools/gradle/endpoints/client/EndpointsClientPlugin.java (hunk 6 above)
Comment (loosebazooka) [reply src/main/java/com/google/cloud/tools/gradle/endpoints/client/EndpointsClientPlugin.java:172]: one is the where the zips go, and one is a src directory that is registered with gradle

File: appveyor.yml (hunk 2 above)
Comment (patflynn) [reply appveyor.yml:1]: does appveyor support linux?

File: src/main/java/com/google/cloud/tools/gradle/endpoints/client/EndpointsClientPlugin.java (hunk 4 above)
Comment (patflynn) [reply src/main/java/com/google/cloud/tools/gradle/endpoints/client/EndpointsClientPlugin.java:45]: that's what the doc says.  Seems useful to me.

**Comment by loosebazooka** [comment]
This PR was actually split into 3 parts that depend on eachother. See #1 #2

File: appveyor.yml (hunk 2 above)
Comment (loosebazooka) [reply appveyor.yml:1]: no

File: src/main/java/com/google/cloud/tools/gradle/endpoints/client/EndpointsClientPlugin.java (hunk 4 above)
Comment (loosebazooka) [reply src/main/java/com/google/cloud/tools/gradle/endpoints/client/EndpointsClientPlugin.java:45]: will put in a subsequent change #4

**Review by patflynn — APPROVED** [review]

**Pull request merged**